### PR TITLE
Update WeakEventSource pattern 

### PR DESCRIPTION
**Pull request by igitur**
Update according to changes in  https://github.com/thomaslevesque/WeakEvent
Contains small performance fixes.

And fix some whitespace issues! ;-)

### CODE REVIEW

File: src/QLNet/Patterns/WeakEventSource.cs [hunk 1]
```diff
@@ -40,7 +40,12 @@ public void Raise()
       {
          lock (_handlers)
          {
-            _handlers.RemoveAll(h => !h.Invoke());
+            var failedHandlers = _handlers
+                .ToArray()
```
Comment (amaggiulli) [inline src/QLNet/Patterns/WeakEventSource.cs]: MAJOR Drop 'ToArray' from the middle of the call chain. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

**Comment by amaggiulli** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/QLNet/Patterns/WeakEventSource.cs (hunk 1 above)
Comment (amaggiulli) [reply src/QLNet/Patterns/WeakEventSource.cs]: Why make a copy with ToArray() ?

Comment (igitur) [reply src/QLNet/Patterns/WeakEventSource.cs]: If I remember correctly, it's to ensure event subscriptions can change while events are being raised.

**Comment by igitur** [comment]
Current file at https://github.com/thomaslevesque/WeakEvent/blob/master/WeakEvent/WeakEventSource.cs

File: src/QLNet/Patterns/WeakEventSource.cs [hunk 2]
```diff
@@ -40,7 +40,12 @@ public void Raise()
       {
          lock (_handlers)
          {
-            _handlers.RemoveAll(h => !h.Invoke());
+            var failedHandlers = _handlers
+                                 .ToArray()
```
Comment (amaggiulli) [inline src/QLNet/Patterns/WeakEventSource.cs]: MAJOR Drop 'ToArray' from the middle of the call chain. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

**Comment by igitur** [comment]
Work in progress...

File: src/QLNet/Patterns/WeakEventSource.cs [hunk 3]
```diff
@@ -133,15 +157,143 @@ public bool Invoke()
...
+         private void AddToIndex(Callback singleHandler, int index)
+         {
+            var hashCode = WeakDelegate.GetHashCode(singleHandler);
+            if (_index.ContainsKey(hashCode))
+               _index[hashCode].Add(index);
+            else
+               _index.Add(hashCode, new List<int> { index });
+         }
+
+         private WeakDelegate this[int index]
```
Comment (amaggiulli) [inline src/QLNet/Patterns/WeakEventSource.cs]: MAJOR Remove the unused private property 'this[]'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)

**Pull request closed without merging**